src/openforms/utils/json_logic.py

```diff
@@ -49,3 +52,174 @@ def normalize_value(value: Union[dict, list, tuple, int, str]) -> JSONLogicValue
         return JsonLogicTest.from_expression(value)
 
     raise NotImplementedError(f"Unknown value type: {type(value)}")
+
+
+@dataclass
+class ComponentMeta:
+    form_step: "FormStep"
+    component: Dict[str, Any]
+
+
+@dataclass
+class IntrospectionResult:
+
+    expression: dict
+    rule_tree: Dict[Any, str]
+
+    def as_string(self) -> str:
+        return stringify_json_tree(self.rule_tree)
+
+    def get_input_components(self):
+        return get_leaves(self.rule_tree)
+
+
+def introspect_json_logic(
+    expression: Dict[Any, str],
```

Comment:
your types at not very consistent - I think I'd define at the module level a type:

```py
JsonLogicExpression = Dict[str, Any]
```

you can then reuse `JsonLogicExpression` for the type hints everywhere (instead of `dict` or `Dict[...]`)